core/src/shred_fetch_stage.rs

```diff
@@ -50,26 +50,27 @@ impl ShredFetchStage {
             .as_ref()
             .map(|(_, cluster_info)| cluster_info.keypair().clone());
 
-        // Only need root bank in order to check feature statuses later on;
-        // can demote to only fetching root slot once those features go away.
-        let (mut root_bank, mut last_slot) = {
+        let (mut last_root, mut last_slot, mut slots_per_epoch) = {
             let bank_forks_r = bank_forks.read().unwrap();
-            (bank_forks_r.root_bank(), bank_forks_r.highest_slot())
+            let root_bank = bank_forks_r.root_bank();
+            (
+                root_bank.slot(),
+                root_bank.get_slots_in_epoch(root_bank.epoch()),
+                bank_forks_r.highest_slot(),
+            )
         };
-        let mut last_root = root_bank.slot();
-        let mut slots_per_epoch = root_bank.get_slots_in_epoch(root_bank.epoch());
         let mut stats = ShredFetchStats::default();
 
         for mut packet_batch in recvr {
             if last_updated.elapsed().as_millis() as u64 > DEFAULT_MS_PER_SLOT {
                 last_updated = Instant::now();
                 {
                     let bank_forks_r = bank_forks.read().unwrap();
-                    root_bank = bank_forks_r.root_bank();
+                    let root_bank = bank_forks_r.root_bank();
```

Comment:
Probably does not matter much, but
```rust
let root_bank = bank_forks.read().unwrap().root_bank();
```
wouldn't unnecessarily hold on to bank-forks lock.

Reply:
Yeah, that thought crossed my mind too. However, we also need the lock in order to call `BankForks::highest_slot()`. But, I can tweak things slightly to minimize operations done with lock held

Reply:
Oh, I didn't notice `BankForks::highest_slot`. My bad.
In that case the old code was perfectly fine; no need to make nested scopes (wouldn't really save anything measurable).